DOCS-4289 OAuth 2.0 Support in HTTP Requests Doc Update (#15643)
DataDog/documentation: content/en/synthetics/guide/app-that-requires-login.md

---
title: Running Tests On An Application That Requires Authentication
kind: guide
description: Learn how to ensure your Synthetic browser tests can log in to your applications. 
further_reading:
  - link: 'https://www.datadoghq.com/blog/test-creation-best-practices/'
    tag: 'Blog'
    text: 'Best practices for creating end-to-end tests'
  - link: '/synthetics/guide/browser-tests-totp'
    tag: 'Documentation'
    text: 'TOTPs for Multi-Factor Authentication (MFA) in browser tests'
  - link: '/synthetics/browser_tests'
    tag: 'Documentation'
    text: 'Learn about browser tests'
  - link: '/synthetics/browser_tests/actions'
    tag: 'Documentation'
    text: 'Learn about browser test steps'
---

## Overview 

> <div class="alert alert-info"> Are you interested in testing applications sitting behind MFA? Visit the <a href="/synthetics/guide/app-that-requires-login/#multi-factor-authentication" target="_blank">Multi-factor authentication section</a> and <a href="https://docs.google.com/forms/d/e/1FAIpQLSdjx8PDZ8kJ3MD2ehouTri9z_Fh7PoK90J8arRQgt7QFgFxog/viewform?usp=sf_link">send us feedback</a> to help us work on the systems that matter the most to your teams.</div>

Reviewer: ```suggestion
<div class="alert alert-info">If you are interested in testing applications sitting behind MFA, visit the <a href="/synthetics/guide/app-that-requires-login/#multi-factor-authentication" target="_blank">Multi-factor authentication section</a> and <a href="https://docs.google.com/forms/d/e/1FAIpQLSdjx8PDZ8kJ3MD2ehouTri9z_Fh7PoK90J8arRQgt7QFgFxog/viewform?usp=sf_link">send feedback</a> to help Datadog work on the systems that matter the most to your teams.</div>
```

  - link: '/synthetics/guide/browser-tests-totp'
    tag: 'Documentation'
    text: 'TOTPs for Multi-Factor Authentication (MFA) in browser tests'
  - link: '/synthetics/browser_tests'
    tag: 'Documentation'
    text: 'Learn about browser tests'
  - link: '/synthetics/browser_tests/actions'
    tag: 'Documentation'
    text: 'Learn about browser test steps'
---

## Overview 

<div class="alert alert-info">If you are interested in testing applications sitting behind MFA, visit the <a href="/synthetics/guide/app-that-requires-login/#multi-factor-authentication" target="_blank">Multi-factor authentication section</a> and <a href="https://docs.google.com/forms/d/e/1FAIpQLSdjx8PDZ8kJ3MD2ehouTri9z_Fh7PoK90J8arRQgt7QFgFxog/viewform?usp=sf_link">send feedback</a> to help Datadog work on the systems that matter the most to your teams.</div>

You may need to monitor user journeys located behind a login. There are two ways to ensure that your Datadog browser tests can go through the login steps of your application to perform validation on post-login pages:

- [Include the login steps in your browser test recording](#include-the-login-steps-in-your-recording)

> - [Leverage advanced options for API, multistep, and browser tests](#leverage-test-configuration-options)

Author: ```suggestion
- [Leverage advanced options in your browser tests](#leverage-test-configuration-options)
```

You can also ensure your credentials are securely stored and obfuscated across the application by [using obfuscated global variables](#account-security).

> To ensure your credentials are securely stored and obfuscated across the application, use [secured global variables](#account-security).

Author: ```suggestion
To ensure your credentials are securely stored and obfuscated across the application, use [obfuscated global variables](#account-security).
```

<div class="alert alert-info">Datadog is constantly adding features to help you record test scenarios more easily. <a href="https://docs.google.com/forms/d/e/1FAIpQLSdjx8PDZ8kJ3MD2ehouTri9z_Fh7PoK90J8arRQgt7QFgFxog/viewform?usp=sf_link">Sending feedback</a> about the MFA systems that matter the most to you is appreciated.</div>

## Leverage test configuration options

> The second method is to use one or several of the available advanced configuration options for [HTTP tests][6], [HTTP steps in browser tests][7], and [multistep API tests][8]. 

Reviewer: ```suggestion
The second method is to use one or several of the available advanced configuration options for [HTTP tests][6], [HTTP steps in browser tests][7], or [multistep API tests][8]. 
```

You can apply the following options to your HTTP request:

- Specific request headers
- Cookies

> - Credentials for Basic Auth, Digest Auth, NTLM credentials, AWS Signature, or OAuth 2.0

Reviewer: ```suggestion
- Credentials for basic authentication, digest authentication, NTLM, AWS Signature, or OAuth 2.0
```

These configuration options are set at each test execution and apply to every browser test step at execution time, not recording time. 

{{< img src="synthetics/guide/app_that_requires_login/http_request_advanced_options_oauth.png" alt="Login to your app with browser test configuration options such as OAuth 2.0" style="width:90%" >}}

You can manually apply these configured headers, cookies, and credentials on the page you are recording from and record the steps your test performs post-login. 

By default, the browser, API, or multistep API test automatically passes through authentication with your specified headers, cookies, and credentials at execution time and then completes all recorded steps.

## Secure your account

### Secure your authentication data

Store your credentials as [global variables][9] (for example, one global variable for username, another one for password) and set these variables as secure to obfuscate their values from anyone else who has access to your instance of Datadog.

> Once you have created secure variables, you can [import these global variables][10] into your browser tests and leverage them for your login steps, or use secure variables in your HTTP and multistep API tests.

Reviewer 2: we do not say 'secure variables' anymore but 'hidden' or 'obfuscated' variables

## Leverage test configuration options

The second method is to use one or several of the available advanced configuration options for [HTTP tests][6], [HTTP steps in browser tests][7], or [multistep API tests][8]. 

You can apply the following options to your HTTP request:

- Specific request headers
- Cookies
- Credentials for basic authentication, digest authentication, NTLM, AWS Signature, or OAuth 2.0

These configuration options are set at each test execution and apply to every browser test step at execution time, not recording time. 

{{< img src="synthetics/guide/app_that_requires_login/http_request_advanced_options_oauth.png" alt="Login to your app with browser test configuration options such as OAuth 2.0" style="width:90%" >}}

You can manually apply these configured headers, cookies, and credentials on the page you are recording from and record the steps your test performs post-login. 

> By default, the browser, API, or multistep API test automatically passes through authentication with your specified headers, cookies, and credentials at execution time and then completes all recorded steps.

Author: ```suggestion
By default, the browser test automatically passes through authentication with your specified headers, cookies, and credentials at execution time and then completes all recorded steps.
```
